src/modules/topoh/topoh_mod.c

```diff
@@ -179,82 +192,228 @@ static int mod_init(void)
 			goto error;
 		}
 	}
-	if(th_ip.len<=0)
+
+	if(th_ip.len != 0) {
+		if(th_ip.len + 32 >= MAX_URI_SIZE) {
+			LM_ERR("mask address is too long\n");
+			goto error;
+		}
+		memcpy(buri, "sip:", 4);
+		memcpy(buri+4, th_ip.s, th_ip.len);
+		buri[th_ip.len+8] = '\0';
+
+		if(parse_uri(buri, th_ip.len+4, &puri)<0) {
+			LM_ERR("mask uri is invalid\n");
+			goto error;
+		}
+		if(check_self(&puri.host, puri.port_no, 0)==1)
+		{
+			th_mask_addr_myself = 1;
+			LM_INFO("mask address matches myself [%.*s]\n",
+					th_ip.len, th_ip.s);
+		}
+
+		if(th_build_via_prefix(&th_via_prefix, &th_ip))
+		{
+			goto error;
+		}
+		if(th_build_uri_prefix(&th_uri_prefix, &th_ip))
+		{
+			goto error;
+		}
+	} else {
+		th_socket_hash_table = pkg_malloc(sizeof(struct str_hash_table));
+		if(th_socket_hash_table == NULL){
+			PKG_MEM_ERROR_FMT("th_socket_hash_table\n");
+			goto error;
+		}
+		if(str_hash_alloc(th_socket_hash_table, TH_HT_SIZE))
+			goto error;
+
+		str_hash_init(th_socket_hash_table);
+		if(th_parse_socket_list(*get_sock_info_list(PROTO_UDP)) != 0 ||
+		   th_parse_socket_list(*get_sock_info_list(PROTO_TCP)) != 0 ||
+		   th_parse_socket_list(*get_sock_info_list(PROTO_TLS)) != 0 ||
+		   th_parse_socket_list(*get_sock_info_list(PROTO_SCTP)) !=0)
+			goto error;
```

Comment:
We should probably free the th_socket_hash_table and the str_hash_alloc list in error block.